crates/red_knot_python_semantic/resources/mdtest/shadowing/function.md

```diff
@@ -25,3 +25,27 @@ def f(): ...
 
 f: int = 1
 ```
+
+## Explicit shadowing involving `def` statements
+
+```py path=a.py
+from typing_extensions import reveal_type
+
+f = 1
+reveal_type(f)  # revealed: Literal[1]
+
+def f(): ...
+
+reveal_type(f)  # revealed: Literal[f]
+
+def f(x: int) -> int: ...
+
+reveal_type(f)  # revealed: Literal[f]
+
+f: int = 1
+reveal_type(f)  # revealed: Literal[1]
+
+def f(x: int) -> int: ...
```

Comment:
Including annotations on the second and third definitions of `x` but not the first one is potentially confusing and not really related to the test. Also `-> int: ...` in a non-stub is likely to be a type error in future. So I'd be inclined to remove all the parameter and return annotations here? If we need a separate test to show that presence or absence of annotations doesn't affect whether a `def` statement is a declaration or not, we can add that as a separate test, with explicit commentary.
```suggestion
def f(): ...

reveal_type(f)  # revealed: Literal[f]

f: int = 1
reveal_type(f)  # revealed: Literal[1]

def f(): ...
```

Reply:
I kept one of these with annotations, to show that re-declaring a `def` with a different type is not an error.